File: view/src/schema/reservation.ts
import { Room } from "src/schema/room";
import { ReservationMethod } from "src/schema/reservation-method";
import {
  DynamicRuleMap,
  EnumMap,
  FieldDetail,
  FieldMap,
  StaticRuleMap,
} from "src/types/map";
import { formatDate, formatDateTime, formatPrice } from "src/util/format-util";
import {
  BASE_AUDIT_FIELD_MAP,
  BASE_TIME_FIELD_MAP,
  BaseAudit,
  BaseTime,
} from "src/schema/base";

const RESERVATION_STATUS_MAP: EnumMap = {
  NORMAL: "예약 확정",
  PENDING: "예약 대기",
  CANCEL: "취소 요청",
  REFUND: "환불 완료",
} as const;
export type ReservationStatus = keyof typeof RESERVATION_STATUS_MAP;

export function reservationStatusValueToName(role: ReservationStatus) {
  return RESERVATION_STATUS_MAP[role] || role;
}

export type Reservation = {
  id: number;
  reservationMethod: ReservationMethod;
  rooms: Room[];
  name: string;
  phone: string;
  peopleCount: number;
  stayStartAt: string;
  stayEndAt: string;
  checkInAt: string;
  checkOutAt: string;
  price: number;
  paymentAmount: number;
  refundAmount: number;
  brokerFee: number;
  note: string;
  canceledAt: string;
  status: ReservationStatus;
} & BaseTime &
  BaseAudit;

const ReservationFieldMap: FieldMap = {
  id: { label: "ID" } as const,
  reservationMethod: {
    label: "예약 수단",
    format: (value) => value.name,
  } as const,
  // NOTE: 히스토리 조회 시 하위 호환을 위해 유지 필요
  room: {
    label: "객실",
    format: (value: Room) => (value ? value.number : "미배정"),
  } as const,
  rooms: {
    label: "객실",
    format: (value: Room[]) =>
      value.length !== 0
        ? value.map((room) => room.number).join(", ")
        : "미배정",
  } as const,
  name: { label: "예약자명" } as const,
  phone: {
    label: "예약자 연락처",
    format: (value: string) => value || "-",
  } as const,
  peopleCount: { label: "예약인원" } as const,
  stayStartAt: { label: "입실일", format: formatDate } as const,
  stayEndAt: { label: "퇴실일", format: formatDate } as const,
  checkInAt: {
    label: "체크인 시각",
    format: (value: string | null) => (value ? formatDateTime(value) : ""),
  } as const,
  checkOutAt: {
    label: "체크아웃 시각",
    format: (value: string | null) => (value ? formatDateTime(value) : ""),
  } as const,
  price: { label: "판매 금액", format: formatPrice } as const,
  paymentAmount: { label: "누적 결제 금액", format: formatPrice } as const,
  refundAmount: { label: "환불 금액", format: formatPrice } as const,
  brokerFee: { label: "예약 수단 수수료", format: formatPrice } as const,
  note: { label: "메모" } as const,
  canceledAt: { label: "취소 시각", format: formatDateTime } as const,
  status: { label: "상태", format: reservationStatusValueToName } as const,
  ...BASE_TIME_FIELD_MAP,
  ...BASE_AUDIT_FIELD_MAP,
} as const;

export function getReservationFieldDetail(field: string): FieldDetail | null {
  const fieldDetail = ReservationFieldMap[field];

  return fieldDetail
    ? {
        field: field,
        ...fieldDetail,
      }
    : null;
}

export function formatReservationFieldToLabel(field: string) {
  return getReservationFieldDetail(field)?.label ?? field;
}

export function formatReservationValue(
  field: string,
  value: string | number | null,
) {
  const fieldDetail = getReservationFieldDetail(field);
  return fieldDetail?.format ? fieldDetail.format(value) : value;
}

export const reservationStaticRules: StaticRuleMap = {
  name: [
    (value: string) =>
      (value.length >= 2 && value.length <= 30) || "2~30 글자가 필요합니다",
  ] as const,
  phone: [
    (value: string) => value.length <= 20 || "20 글자 이내로 입력 가능합니다",
  ] as const,
  peopleCount: [
    (value: number) =>
      (value >= 0 && value <= 1000) || "1000 명 이하만 입실 가능합니다",
  ] as const,
  stayStartAt: [
    (value: string) =>
      /^-?[\d]+-[0-1]\d-[0-3]\d$/.test(value) ||
      "####-##-## 형태의 날짜만 입력 가능합니다.",
  ] as const,
  stayEndAt: [
    (value: string) =>
      /^-?[\d]+-[0-1]\d-[0-3]\d$/.test(value) ||
      "####-##-## 형태의 날짜만 입력 가능합니다.",
  ] as const,
  price: [
    (value: number) =>
      (value >= 0 && value <= 100000000) || "금액은 1억 미만 양수만 가능합니다",
  ] as const,
  brokerFee: [
    (value: number) =>
      (value >= 0 && value <= 100000000) || "금액은 1억 미만 양수만 가능합니다",
  ] as const,
  note: [
    (value: string) =>
      (value.length >= 0 && value.length <= 200) ||
      "200 글자까지 입력 가능합니다",
  ] as const,
} as const;

export const reservationDynamicRules: DynamicRuleMap = {
  paymentAmount: (price: number) =>
    [
      (value: number) =>
        (value >= 0 && value <= 100000000) ||
        "금액은 1억 미만 양수만 가능합니다",
      (value: number) => value <= price || "판매 금액보다 클 수 없습니다",
    ] as const,
} as const;

File: apps/api-core/internal/models/reservation.go
package models

import (
	"database/sql/driver"
	"sort"
	"time"

	"gorm.io/gorm"
)

type ReservationStatus int8

const (
	ReservationStatusRefund  ReservationStatus = -10
	ReservationStatusCancel  ReservationStatus = -1
	ReservationStatusPending ReservationStatus = 0
	ReservationStatusNormal  ReservationStatus = 1
)

func (s ReservationStatus) String() string {
	switch s {
	case ReservationStatusRefund:
		return "REFUND"
	case ReservationStatusCancel:
		return "CANCEL"
	case ReservationStatusPending:
		return "PENDING"
	case ReservationStatusNormal:
		return "NORMAL"
	default:
		return "UNKNOWN"
	}
}

func (s ReservationStatus) Value() (driver.Value, error) {
	return int64(s), nil
}

func (s *ReservationStatus) Scan(value interface{}) error {
	if value == nil {
		*s = ReservationStatusPending
		return nil
	}
	switch v := value.(type) {
	case int64:
		*s = ReservationStatus(v)
	case int8:
		*s = ReservationStatus(v)
	default:
		*s = ReservationStatusPending
	}
	return nil
}

type ReservationType int8

const (
	ReservationTypeStay        ReservationType = 0
	ReservationTypeMonthlyRent ReservationType = 10
)

func (t ReservationType) String() string {
	switch t {
	case ReservationTypeStay:
		return "STAY"
	case ReservationTypeMonthlyRent:
		return "MONTHLY_RENT"
	default:
		return "UNKNOWN"
	}
}

func (t ReservationType) Value() (driver.Value, error) {
	return int64(t), nil
}

func (t *ReservationType) Scan(value interface{}) error {
	if value == nil {
		*t = ReservationTypeStay
		return nil
	}
	switch v := value.(type) {
	case int64:
		*t = ReservationType(v)
	case int8:
		*t = ReservationType(v)
	default:
		*t = ReservationTypeStay
	}
	return nil
}

type Reservation struct {
	BaseMustAuditEntity
	PaymentMethodID uint              `gorm:"column:payment_method_id;not null" json:"paymentMethodId"`
	PaymentMethod   *PaymentMethod    `gorm:"foreignKey:PaymentMethodID" json:"paymentMethod,omitempty"`
	Rooms           []ReservationRoom `gorm:"foreignKey:ReservationID" json:"rooms,omitempty"`
	Name            string            `gorm:"column:name;type:varchar(30);not null" json:"name"`
	Phone           string            `gorm:"column:phone;type:varchar(15);not null" json:"phone"`
	PeopleCount     int               `gorm:"column:people_count;not null;default:0" json:"peopleCount"`
	StayStartAt     time.Time         `gorm:"column:stay_start_at;type:date;not null" json:"stayStartAt"`
	StayEndAt       time.Time         `gorm:"column:stay_end_at;type:date;not null" json:"stayEndAt"`
	CheckInAt       *time.Time        `gorm:"column:check_in_at;type:datetime" json:"checkInAt,omitempty"`
	CheckOutAt      *time.Time        `gorm:"column:check_out_at;type:datetime" json:"checkOutAt,omitempty"`
	Price           int               `gorm:"not null" json:"price"`
	Deposit         int               `gorm:"not null;default:0" json:"deposit"`
	PaymentAmount   int               `gorm:"column:payment_amount;not null;default:0" json:"paymentAmount"`
	RefundAmount    int               `gorm:"column:refund_amount;not null;default:0" json:"refundAmount"`
	BrokerFee       int               `gorm:"column:broker_fee;not null;default:0" json:"brokerFee"`
	Note            string            `gorm:"type:varchar(200)" json:"note"`
	CanceledAt      *time.Time        `gorm:"column:canceled_at" json:"canceledAt,omitempty"`
	Status          ReservationStatus `gorm:"type:tinyint;not null;default:0" json:"status"`
	Type            ReservationType   `gorm:"type:tinyint;not null;default:0" json:"type"`
}

func (Reservation) TableName() string {
	return "reservation"
}

func (r *Reservation) BeforeCreate(tx *gorm.DB) error {
	if err := r.BaseMustAuditEntity.BeforeCreate(tx); err != nil {
		return err
	}
	if r.Status == 0 {
		r.Status = ReservationStatusPending
	}
	if r.Type == 0 {
		r.Type = ReservationTypeStay
	}
	if r.Note == "" {
		r.Note = ""
	}
	return nil
}

func (r *Reservation) IsActive() bool {
	return r.Status == ReservationStatusNormal || r.Status == ReservationStatusPending
}

func (r *Reservation) IsCanceled() bool {
	return r.Status == ReservationStatusCancel || r.Status == ReservationStatusRefund
}

func (r *Reservation) GetStayDays() int {
	return int(r.StayEndAt.Sub(r.StayStartAt).Hours() / 24)
}

// GetAuditEntityType implements audit.Auditable interface
func (r *Reservation) GetAuditEntityType() string {
	return "reservation"
}

// GetAuditEntityID implements audit.Auditable interface
func (r *Reservation) GetAuditEntityID() uint {
	return r.ID
}

func formatTimePtr(t *time.Time) interface{} {
	if t == nil {
		return nil
	}
	return t.Format(time.RFC3339)
}

// GetAuditFields implements audit.Auditable interface
func (r *Reservation) GetAuditFields() map[string]interface{} {
	rooms := make([]map[string]interface{}, 0)
	for _, rr := range r.Rooms {
		roomData := map[string]interface{}{
			"id":     rr.RoomID,
			"number": "",
		}
		if rr.Room != nil {
			roomData["number"] = rr.Room.Number
		}
		rooms = append(rooms, roomData)
	}
	sort.Slice(rooms, func(i, j int) bool {
		return rooms[i]["id"].(uint) < rooms[j]["id"].(uint)
	})

	paymentMethod := map[string]interface{}{
		"id":   r.PaymentMethodID,
		"name": "",
	}
	if r.PaymentMethod != nil {
		paymentMethod["name"] = r.PaymentMethod.Name
	}

	return map[string]interface{}{
		"id":            r.ID,
		"rooms":         rooms,
		"paymentMethod": paymentMethod,
		"name":          r.Name,
		"phone":         r.Phone,
		"peopleCount":   r.PeopleCount,
		"stayStartAt":   r.StayStartAt.Format("2006-01-02"),
		"stayEndAt":     r.StayEndAt.Format("2006-01-02"),
		"checkInAt":     formatTimePtr(r.CheckInAt),
		"checkOutAt":    formatTimePtr(r.CheckOutAt),
		"price":         r.Price,
		"deposit":       r.Deposit,
		"paymentAmount": r.PaymentAmount,
		"refundAmount":  r.RefundAmount,
		"brokerFee":     r.BrokerFee,
		"note":          r.Note,
		"canceledAt":    formatTimePtr(r.CanceledAt),
		"status":        r.Status.String(),
		"type":          r.Type.String(),
		"createdBy":     r.CreatedBy,
		"updatedBy":     r.UpdatedBy,
		"createdAt":     r.CreatedAt,
		"updatedAt":     r.UpdatedAt,
	}
}
